wrangler-api/src/main/java/io/cdap/wrangler/api/Directive.java

```diff
@@ -126,4 +132,20 @@ default List<EntityCountMetric> getCountMetrics() {
     // no op
     return null;
   }
+
+  /**
+   * This method provides a way to specify the output {@link Schema} generated by a directive after it performs its
+   * transformation on the data.
+   * @param inputSchema schema of the input data
+   * @param generator {@link SchemaFieldGenerator} which can generate a {@link Schema.Field}
+   *                                              given a field name and value (Java object)
+   * @param output output data generated by directive after its execution
+   * @return {@link Schema} of the output data
+   * @implNote By default, the implementation returns a null (no-op)
+   * */
+  @Override
+  default Schema getOutputSchema(Schema inputSchema, SchemaFieldGenerator generator, List<Row> output) {
```

Comment:
Mark as Nullable. 
Also, why would a Directive need params `generator` and `output`?

Reply:
Overall I think we can remove most of the generator changes as those are now internal to the Wrangler itself

Reply:
UDDs need a way to generate schema in the same way as Wrangler internally does. Ex: can use this interface in custom implementation of the `getOutputSchema` method to say "generate schema for column A but carry over schema for column B". The internal implementation of schema generation logic can be reused by UDD developers as well

Reply:
Output is needed to generate schema for new columns

Reply:
Changed the `SchemaUpdater` interface to have corresponding methods for all four operations (add, remove, rename and change type). Removed the generator interface and reverted changes to SchemaConverter